File: src/js/sort_selection_sort.js
var fs = require("fs");

// business logic start //

// selection sort - repeatedly pick the smallest remaining element into the output
// https://en.wikipedia.org/wiki/Selection_sort

function doSorting(inputList) {
  return selectionSort(inputList);
}

function selectionSort(inputList) {
  sortedElements = [];
  unsortedElements = inputList;
  inputListLength = inputList.length;

  for (let index = 0; index < inputListLength; index++) {
    smallestIndex = findSmallestIndex(unsortedElements);
    smallestElement = unsortedElements[smallestIndex];
    sortedElements.push(smallestElement);
    unsortedElements.splice(smallestIndex, 1);
  }

  return sortedElements;
}

function findSmallestIndex(inputList) {
  smallestIndex = 0;

  inputList.forEach((element, index) => {
    if (element < inputList[smallestIndex]) {
      smallestIndex = index;
    }
  });

  return smallestIndex;
}

// business logic end //

// 👇🏽 copy pasted helpers

// read input file
let inputFilePath = process.env.INPUT_PATH;
let inputString = fs.readFileSync(inputFilePath, "utf8");
let inputStringSplit = inputString.split(/\n/);

// clean input data
if (inputStringSplit[inputStringSplit.length - 1] === "") {
  inputStringSplit = inputStringSplit.slice(0, inputStringSplit.length - 1);
}

// do sorting
sortedData = doSorting(inputStringSplit);

// write output file
let outputFilePath = process.env.OUTPUT_PATH;
let outputString = sortedData.join("\n");
outputString += "\n"; // trailing newline
fs.writeFileSync(outputFilePath, outputString);

File: src/ruby/sort_selection_sort.rb
# frozen_string_literal: true

# business logic start #

# selection sort - https://en.wikipedia.org/wiki/Selection_sort
# Repeatedly find the smallest unsorted element and append it to the output list.

def do_sorting(input_list)
  selection_sort(input_list)
end

def selection_sort(input_list)
  sorted_elements = []
  unsorted_elements = input_list

  (0..unsorted_elements.length).each do |_|
    smallest_index = find_smallest_index(unsorted_elements)
    smallest_element = unsorted_elements[smallest_index]
    sorted_elements.append(smallest_element)
    # ruby `delete_at` is the equivalent of python `pop` (ruby `pop` differs)
    unsorted_elements.delete_at(smallest_index)
  end

  sorted_elements
end

def find_smallest_index(input_list)
  smallest_index = 0

  input_list.each_with_index do |element, index|
    smallest_index = index if element < input_list[smallest_index]
  end

  smallest_index
end

# business logic end #

if __FILE__ == $PROGRAM_NAME
  input_file_path = ENV['INPUT_PATH']
  input_file = File.readlines(input_file_path)

  sorted_data = do_sorting(input_file)

  output_file_path = ENV['OUTPUT_PATH']
  File.open(output_file_path, 'wb') { |f| f.write(sorted_data.join('')) }
end
